kasa/smart/modules/firmware.py

```diff
@@ -85,23 +86,34 @@ def query(self) -> dict:
     def latest_firmware(self):
         """Return latest firmware information."""
         fw = self.data.get("get_latest_fw") or self.data
-        if isinstance(fw, SmartErrorCode):
+        if not self._device._is_cloud_connected or isinstance(fw, SmartErrorCode):
             # Error in response, probably disconnected from the cloud.
             return UpdateInfo(type=0, need_to_upgrade=False)
 
         return UpdateInfo.parse_obj(fw)
 
     @property
-    def update_available(self):
+    def update_available(self) -> bool | None:
         """Return True if update is available."""
+        if not self._device._is_cloud_connected:
+            return None
         return self.latest_firmware.update_available
 
     async def get_update_state(self):
         """Return update state."""
+        if not self._device._is_cloud_connected:
+            raise KasaException(
+                "Device must be connected to the internet "
+                "to get firmware download state."
+            )
         return await self.call("get_fw_download_state")
 
     async def update(self):
         """Update the device firmware."""
+        if not self._device._is_cloud_connected:
```

Comment:
```suggestion
        if not self._device.is_cloud_connected:
```
I'm still wondering if we should keep this simple and just let users to call the methods that can fail? We need to perform the check for the exposed features, but we could skip it for actions, IMO.

Reply:
Fixed